File: sdk/src/capybara/_schema.py
"""Tool output schema registry: Pydantic models serialized onto spans."""

from __future__ import annotations

import json
from typing import TYPE_CHECKING

from opentelemetry.sdk.trace import SpanProcessor

if TYPE_CHECKING:
    from opentelemetry.context import Context
    from opentelemetry.sdk.trace import Span
    from pydantic import BaseModel

SCHEMA_ATTR = "capybara.schema"
_TOOL_NAME_ATTRS = ("gen_ai.tool.name", "mcp.tool.name")

_registry: dict[str, str] = {}


def schema(tool_name: str, model: type[BaseModel]) -> type[BaseModel]:
    """Register a tool's Pydantic output model; attached to matching tool spans."""
    _registry[tool_name] = json.dumps(model.model_json_schema(), separators=(",", ":"))
    return model


def schema_for(tool_name: str) -> str | None:
    return _registry.get(tool_name)


# SchemaSpanProcessor attaches declared schemas to third-party tool spans whose
# tool name is already set at span start; capybara.trace attaches its own.
class SchemaSpanProcessor(SpanProcessor):
    def on_start(self, span: Span, parent_context: Context | None = None) -> None:
        attributes = span.attributes or {}
        for key in _TOOL_NAME_ATTRS:
            name = attributes.get(key)
            if isinstance(name, str) and name in _registry:
                span.set_attribute(SCHEMA_ATTR, _registry[name])
                return

File: sdk-js/src/schema.ts
// Tool output schema registry: JSON Schemas attached to matching tool spans.
import type { Span, SpanProcessor, ReadableSpan } from '@opentelemetry/sdk-trace-base';

export const SCHEMA_ATTR = 'capybara.schema';
const TOOL_NAME_ATTRS = ['gen_ai.tool.name', 'mcp.tool.name'];

const registry = new Map<string, string>();

// schema registers a tool's declared output shape. capybara understands the
// JSON Schema subset {type, properties, required, items}, so a plain object
// works and no schema library is pulled in.
export function schema<T>(toolName: string, jsonSchema: T): T {
  registry.set(toolName, typeof jsonSchema === 'string' ? jsonSchema : JSON.stringify(jsonSchema));
  return jsonSchema;
}

export function schemaFor(toolName: string): string | undefined {
  return registry.get(toolName);
}

// SchemaSpanProcessor attaches a declared schema to a third-party tool span
// whose tool name is already set at span start; trace() attaches its own.
export class SchemaSpanProcessor implements SpanProcessor {
  onStart(span: Span): void {
    const attrs = (span as unknown as { attributes: Record<string, unknown> }).attributes ?? {};
    for (const key of TOOL_NAME_ATTRS) {
      const name = attrs[key];
      if (typeof name === 'string') {
        const declared = registry.get(name);
        if (declared !== undefined) {
          span.setAttribute(SCHEMA_ATTR, declared);
          return;
        }
      }
    }
  }
  onEnd(_span: ReadableSpan): void {}
  shutdown(): Promise<void> {
    return Promise.resolve();
  }
  forceFlush(): Promise<void> {
    return Promise.resolve();
  }
}
